2.0/m_extbanregex.cpp

```diff
@@ -0,0 +1,362 @@
+
+	for (unsigned int i = 0; i < masks.size(); ++i)
+	{
+		// Send the mode line and start over before it gets truncated
+		if (strlength > 450 || modestr[1].length() > 50)
+		{
+			ServerInstance->SendMode(modestr, ServerInstance->FakeClient);
+			modestr.resize(1);
+			modestr.push_back("-");
+			strlength = 7 + chan.length();
+		}
+		modestr[1].push_back(modechar);
+		modestr.push_back(masks[i]);
+		strlength += 2 + masks[i].length();
+	}
+
+	ServerInstance->SendMode(modestr, ServerInstance->FakeClient);
+}
+
+void RemoveAll(const std::string& engine, bool runexceptions, bool runinvex)
+{
+	ListModeBase* banexception = NULL;
+	ListModeBase* invex = NULL;
+
+	// We know these modes are proper if their ModeWatcher is active
+	if (runexceptions)
+		banexception = (ListModeBase*)ServerInstance->Modes->FindMode('e', MODETYPE_CHANNEL);
+	if (runinvex)
+		invex = (ListModeBase*)ServerInstance->Modes->FindMode('I', MODETYPE_CHANNEL);
+
+	/* Loop each channel checking for any regex extbans
+	 * Create a list to remove and batch the mode removal
+	 * Send a notice to hop/op if anything was removed
+	 */
+	const chan_hash& chans = *ServerInstance->chanlist;
+	for (chan_hash::const_iterator i = chans.begin(); i != chans.end(); ++i)
+	{
+		bool removed = false;
+		Channel* chan = i->second;
+
+		std::vector<std::string> banmasks;
+		for (BanList::iterator ban = chan->bans.begin(); ban != chan->bans.end(); ++ban)
+		{
+			if (IsExtBanRegex(ban->data))
+				banmasks.push_back(ban->data);
+		}
+		if (!banmasks.empty())
+		{
+			ModeLine(chan->name, 'b', banmasks);
+			removed = true;
+		}
+
+		modelist* banexceptions = banexception ? banexception->extItem.get(chan) : NULL;
+		if (banexceptions)
+		{
+			std::vector<std::string> excmasks;
+			for (modelist::iterator exc = banexceptions->begin(); exc != banexceptions->end(); ++exc)
+			{
+				if (IsExtBanRegex(exc->mask))
+					excmasks.push_back(exc->mask);
+			}
+			if (!excmasks.empty())
+			{
+				ModeLine(chan->name, 'e', excmasks);
+				removed = true;
+			}
+		}
+
+		modelist* inviteexceptions = invex ? invex->extItem.get(chan) : NULL;
+		if (inviteexceptions)
+		{
+			std::vector<std::string> invmasks;
+			for (modelist::iterator inv = inviteexceptions->begin(); inv != inviteexceptions->end(); ++inv)
+			{
+				if (IsExtBanRegex(inv->mask))
+					invmasks.push_back(inv->mask);
+			}
+			if (!invmasks.empty())
+			{
+				ModeLine(chan->name, 'I', invmasks);
+				removed = true;
+			}
+		}
+
+		if (!removed)
+			continue;
+
+		CUList empty;
+		std::string notice = "Regex engine has changed to '" + engine + "'. All regex extbans have been removed";
+		ModeHandler* hop = ServerInstance->Modes->FindMode('h', MODETYPE_CHANNEL);
+		char pfxchar = ((hop) && (hop->name == "halfop")) ? hop->GetPrefix() : '@';
+
+		chan->WriteAllExcept(ServerInstance->FakeClient, true, pfxchar, empty, "NOTICE %s :%s", chan->name.c_str(), notice.c_str());
+	}
+}
+
+class BanWatcher : public ModeWatcher
```

Comment:
Can all of these be made into one class which takes the mode to watch as a parameter?

Reply:
Totally never thought to do it that way, should work.